### ISSUE TITLE
Issue with GPU

### ISSUE THREAD

**Issue by wuyao3561** [abTEM/abTEM#72]
I am trying to work with GPU but always encounter an OutOfMemoryError, even if I assign the storage to be CPU. I tried on a workstation with 12G GPU and some large models also not work. 

By the way, would you develop a version that could run using multiple GPU? That would be great to reduce the simulation time.

**Comment by sdfunni** [comment]
I have been testing out abtem over the last few days and have run in to similar problems as well. They seem to occur when I use a potential pixel count  for the slices (the "gpts" argument in Potentials) higher than ~2k x 2k for multislice simulations. I also get out of memory errors whenever running PRISM simulations, almost regardless of size. I have the most  recent beta version but am testing out the latest non-beta version available here (1.0.022). I am using a Nvidia Quadro RTX 4000 w/ 8 GB of gpu memory. Not sure if the potential pixel count represents a fundamental limit to calculation size for the gpu memory.

**Comment by jacobjma** [comment]
It is certainly possible to run out of memory. Could you provide more info? Preferably as reproducible code.

Do these large examples run on your computer? They run on my GPU with 11GB.
https://github.com/jacobjma/abTEM/blob/master/examples/benchmarks/cbed_8_million_atoms.ipynb
https://github.com/jacobjma/abTEM/blob/master/examples/benchmarks/1_million_probe_positions.ipynb

> By the way, would you develop a version that could run using multiple GPU? That would be great to reduce the simulation time.

We have an early version of abTEM using Dask, so multi-GPU support should be available in the near future, see #61. You are very welcome to give the branch a try already now and report your experience.

**Comment by wuyao3561** [comment]
I just paste my code below:

```python

from abtem import __version__

print('current version:', __version__)

from ase.build import mx2
from ase.build import surface
from ase.io import read

from abtem import *
from abtem.structures import orthogonalize_cell

device = 'gpu'
storage='cpu'

##### Read the structure name #####
structure_name=input('Input the structure name: ')

##### Set up atomic structure #####
atoms = read(structure_name+'.cfg')

repetitions = (1, 1, 1)

atoms = orthogonalize_cell(atoms)

atoms *= repetitions

atoms.center(vacuum=2, axis=2)

sigmas = {'Mo': .1, 'W': .1, 'S': .1, 'Se': .1, 'Te': .1} # standard deviations of thermal vibrations
num_configs = 30 # number of frozen phonon configurations

frozen_phonons = FrozenPhonons(atoms, num_configs=num_configs, sigmas=sigmas)
potential = Potential(frozen_phonons, gpts=512, slice_thickness=0.5, parametrization='kirkland', projection='infinite', device=device, storage=storage)

probe = SMatrix(energy=200e3,
                semiangle_cutoff=21,
                expansion_cutoff=25,  # Should be bigger than or equal to the semiangle_cutoff
                rolloff=0.1,
                defocus=0,
                Cs=0,
                focal_spread=0,
                device=device,
                storage=storage)

##### Define detector #####
detector = FlexibleAnnularDetector(save_file=structure_name+'.hdf5')

##### Define scan region #####
end = (potential.extent[0] / repetitions[0], potential.extent[1] / repetitions[1])

gridscan = GridScan(start=[0, 0], end=end, sampling=probe.ctf.nyquist_sampling * .9)

##### Run simulation #####
measurements = probe.scan(gridscan, [detector], potential)

```

**Comment by wuyao3561** [comment]
> It is certainly possible to run out of memory. Could you provide more info? Preferably as reproducible code.
> 
> Do these large examples run on your computer? They run on my GPU with 11GB. https://github.com/jacobjma/abTEM/blob/master/examples/benchmarks/cbed_8_million_atoms.ipynb https://github.com/jacobjma/abTEM/blob/master/examples/benchmarks/1_million_probe_positions.ipynb
> 
> > By the way, would you develop a version that could run using multiple GPU? That would be great to reduce the simulation time.
> 
> We have an early version of abTEM using Dask, so multi-GPU support should be available in the near future, see #61. You are very welcome to give the branch a try already now and report your experience.

Oh i would try with early Dask version. That's great!

**Comment by jacobjma** [comment]
@wuyao3561 To reproduce I need to know the structure as well since the number of plane waves in PRISM depends on the lateral size of the unit cell size.

A quick fix could be manually lowering the batch size. Large batches are faster but require more memory.
```
measurements = probe.scan(gridscan, 
                          [detector], 
                          potential,
                          max_batch_expansion=100, # number of plane waves propagated in parallel
                          max_batch_probes=500) # number of probes reduced in parallel
```
These are usually automatically selected to fit in memory using these functions:
```
print(probe.build()._max_batch_expansion())
print(probe.build()._max_batch_probes())
```
If the estimated values are too large on your system, I would be very interested in understanding why.

One last thing you can try is disabling the FFT plan cache. I found that this can take up a surprising amount of memory. You need to add this at the top of your code:
```
from cupy.fft.config import get_plan_cache
cache = get_plan_cache()
cache.set_size(0)
```

**Comment by wuyao3561** [comment]
[PtTe121.txt](https://github.com/jacobjma/abTEM/files/8079877/PtTe121.txt)
[MoS-3R-4.txt](https://github.com/jacobjma/abTEM/files/8079878/MoS-3R-4.txt)
4-layer MoS2 with 5×5 cell cannot work on my computer with 4G GPU. And I tried on a 12G GPU workstation the PtTe-121 cannot work. i have attached the model in txt form here. To test you could change it to cfg file. Thank you!

**Comment by jacobjma** [comment]
Alright, thanks for attaching the structures. The PtTe-121 runs fine on my 11GB 2080Ti, it finishes in 160s, so this will be a bit harder to debug. I will note, the scattering matrix requires more than 3500 plane waves, so it is not a tiny simulation.

You can very easily get away with setting ``interpolation=2``. I tried doing that, which cut the number of plane waves to 909. This simulation finishes in just 15s, and the result is almost indistinguishable from standard multislice.

Did you try the other things I suggested? In particular running:
```
print(probe.build()._max_batch_expansion())
print(probe.build()._max_batch_probes())
```
What values do you get?

If you know. Are you running out of memory in the multislice step, or the scanning/reduction step?

**Comment by wuyao3561** [comment]
Thank you so much for your suggestion. I would try with the interpolation setting.

I tried with print(probe.build()._max_batch_expansion()), but it required grid extent. I use extent=1 and gpts=512. My laptop with 4G GPU get result of 1 and 327 and workstation with 12G GPU get 1 and 1024.

The multislice step is fine on both my laptop and workstation, but it runs out of memory once the scanning step begins. 

I was also confused about the storage parameter. since I set the storage to CPU do I still require a large GPU memory?
